docs/dynamic-component-loading.md

```diff
@@ -0,0 +1,117 @@
+* `enableUnloading` which will mark the newly create load context to track references and to trigger unload fo the load context and all the assemblies in it when possible.
+
+## High-level description of the implementation
+* Implement a new `AssemblyLoadContext` which will provide the isolation boundary and act as the "root" for the component. It can be enabled for unloading.  
+* The new load context is initialized by specifying the full path to the main assembly of the component to load.  
+* It will look for the `.deps.json` next to that assembly to determine its dependencies. Lack of `.deps.json` will be treated in the same way it is today for executable apps - that is all the assemblies in the same folder will be used as dependencies.
+* Parsing and understanding of the `.deps.json` will be performed by the same host components which do this for executable apps (so same behavior/quirks/bugs, very little code duplication).
+* If the component has `.runtimeconfig.json` and/or `.runtimeconfig.dev.json` it will only be used to verify runtime version and provide probing paths.
+* The load context will determine a list of assemblies similar to the TPA and lists of resources and native search paths and remember them. These will be used to resolve binding events in the load context.
+
+## Handling of various asset types
+What happens with various asset types once the ALC decides to load it in isolation.
+* Normal managed assembly (code) - The `.deps.json` parsing code will return list of full file paths. The ALC will just find it there and load it.  
+Note that R2R images are handled by this as well since they are basically just a slightly different managed assembly. Also note that .NET Core can only load a given R2R image once as R2R. Any subsequent load of the same file will work, but it will be only read as a pure IL assembly (and thus require JITing). Loading the same file multiple times can occur if two load contexts decide to load the assembly in isolation.
+* Satellite assemblies (resources) - Two possibilities:
+    * Imitate app behavior exactly - `.deps.json` only provides list of resource probing paths. ALC would then try to find the `<culture>/AssemblyName.dll` in each probing path and resolve the first match.
+    * Use full paths - `.deps.json` resolution actually internally produces a list of full file paths and then trims it to just probing paths. The full file paths could be used by the ALC in a very similar manner to code assemblies.
+* Native libraries - To integrate well the ALC would only get list of native probing paths from the `.deps.json`. It would then use new API to load native library given a probing path and a simple name. Internally this would call into the existing runtime behavior (which tries various prefix/suffix combinations and so on).
+
+## Isolation strategies
```

Comment:
I like these strategy options - smells like an `enum` instead of a `bool` to me. Could also be a `LoadIsolationPolicy` struct.

Reply:
Are you thinking of different kinds of isolation settings? For example:
* Load everything in the new ALC
* Load everything in the parent/default ALC (in which case, there is no point creating a new ALC)
* Load in the new ALC only if the parent cannot load it
* Load in the new ALC only if the parent cannot load it without using roll-forward
* ...

I think it is best to enable and leave these isolation settings to the developer, rather than providing API support for all of them -- at least to start with.

Reply:
Well the strategies are implying settings, not me. My point here is that a `bool` by its very nature is a binary idea, but based on all these talks it isn't a binary idea rather it is a spectrum. This implies a `bool` is not the appropriate idiom to use for isolation because it isn't "yes or no", but rather "isolate like this ...". An `enum` is one option and a "policy" type struct that indicate how to isolate is another. We are getting into the very concern you and I spoke about earlier this week which is why I am advocating for a very simply scenario that covers a majority of cases and then if a scenario exists that is distinct from the default the developer or consumer of this API will have to answer a series of questions and make real decisions.

Reply:
Just curious which bool are you referring to? The public API as proposed above doesn't provide any way to customize the policy. We've been discussing this quite a bit and the current thinking has somewhat moved, but I'm still trying to prototype some of these ideas before writing it all down. The gist of it is:
- Provide some kind of building blocks, so that devs can compose the policy they want
- If we find a sensible default, implement that as a easy-to-use helper (or possibly the entry point into this problem space) - but this has not been agreed on yet.